File: public/dsa/arrays/pascals_triangle/combinatorics_optimal/pascals_triangle.cpp

#include <bits/stdc++.h>
using namespace std;

vector<int> pascalTriangleRow(int rowNo) {
    int n = rowNo - 1;
    vector<int> pascalRow(n + 1);
    pascalRow[0] = 1;
    for (int k = 1; k <= n; k++) {
        pascalRow[k] = pascalRow[k - 1] * (n - k + 1) / k;
    }
    return pascalRow;
}

// tests-start
int main() {
    int rowNo = 5;
    auto row = pascalTriangleRow(rowNo);
    for (int v : row) {
        cout << v << " ";
    }
    cout << endl;
    return 0;
}
// tests-end

/*output
1 4 6 4 1
*/

File: public/dsa/arrays/pascals_triangle/combinatorics_optimal/pascals_triangle.py


from typing import List


class Solution:
    def pascalTriangleRow(self, rowNo: int) -> List[int]:
        n = rowNo - 1
        result = [0] * (n + 1)
        result[0] = 1
        for k in range(1, n + 1):
            result[k] = result[k-1] * (n - k + 1) // k
        return result


# tests-start
rowNo = 5
sol = Solution()
print(sol.pascalTriangleRow(rowNo))
# tests-end

'''output
[1, 4, 6, 4, 1]
'''
